### ISSUE TITLE
Web Shell - Terminal log for auditing purposes

### ISSUE THREAD

**Issue by marcelofabricanti** [argoproj/argo-cd#9918]
# Summary

In v2.4.0 version, ArgoCD brought us a much waited (and awesome) feature that enables you to shell into containers.
Now developers can have access to their pods freely.

# Motivation

Grant access to people dig into their pods in production environments led us to a few concerns about security and accountability.
Perhaps we need to have visibility about the commands inputed into the pods, since now much more people (from another teams) will have access to do it.

# Proposal

I think would be nice to have a feature that saves an **audit log** on each command executed in web shell feature.

Suggestion of fields:
Timestamp / User Id / argocd project name / argocd app name / cluster of pod / namespace of pod / name of pod / command executed

I'm not sure if store the audit logs somewhere inside the cluster is the best choice.
Or even if store (and make possible search on it afterwards) is a responsability inside Argocd project objectives.

Maybe it's simpler to just export these logs to somewhere else like an AWS S3 Bucket or post to a queue (AWS SQS).
Anyone have a better idea to accomplish this?

**Comment by crenshaw-dev** [comment]
Adding good-first-issue labels because I think the most naiive implementation is very easy - add an env var to the API server to enable logging commands and then add the log statement.

I think logging is a fine first choice rather than another service. If someone has `exec` access, they are already somewhat trusted. And if they want to fill up logs, they probably have other ways to do so.

**Comment by TheSwarnim** [comment]
Hi @crenshaw-dev, please assign this issue to me. I would like to work on it.

**Comment by crenshaw-dev** [comment]
Yep, I forget that this is a stream of characters rather than lines.

**Comment by FuyuanChu** [comment]
Upvote
Could be a killer feature if we could have detailed inputs logged. Complete the missing part of the cluster audit log.
Maybe we can save the detailed inputs and outputs per session.

**Comment by eucleciojosias** [comment]
I would love this feature, I'm struggling to config the SSO with Teleport to make the audit possible

**Comment by crenshaw-dev** [comment]
I think we have to solve the problem of "how do we know what is a command" before we can implement this feature. Logging a series of keystrokes and then backspaces seems maybe worse than no logs at all.

**Comment by Hronom** [comment]
Can we record some how when user hit enter?

**Comment by crenshaw-dev** [comment]
I think we'd want to maintain a cache of keystrokes, applying backspace characters as they arrive. We'd want to flush to logs at a certain cache size limit to avoid filling up memory.

**Comment by Hronom** [comment]
@crenshaw-dev this makes sense too, I think what you described is works. Even if it corrupt line endings and brings two records in logs, I think it will be something instead of nothing

**Comment by atishaninov-xometry** [comment]
I will bump this abandoned discussion with a suggestion.
There is a project [asciinema](https://github.com/asciinema/asciinema) that tries to solve this very problem: record the entire terminal session, including all input, output, and timestamps of specific events, in a compact text file records.

It uses the [asciicast](https://docs.asciinema.org/manual/asciicast/v2/) data storage format, which is actually newline-delimited JSON.

This format solves several logging problems:
1. It doesn't try to interpret input on a line-by-line basis, and so will correctly display sessions with something like vim/nano running in the console
2. It allows data to be logged as soon as it arrives at the terminal, allowing you to guard against browser tab closures and other connection problems
3. It's still a relatively human-readable storage format that can be read when needed without the need for separate players.

To achieve the goals of auditing terminal usage in ArgoCD, we can write such a log to PVC files for each terminal session, potentially encoding some fields in the file path. 
In this case, the choice of storage to store the data is left up to the administrator: it could be some block device, a shared NFS disk, an S3-compatible CSI, or something else that has a Container Storage Interface. And with these files in the PVC, the administrator can already configure separate processes to transfer these logs to the SIEMs if needed, “you figure it out”.

There are two options for the auditor to reproduce and review these records:
1. The easiest and quickest option is to leave access to the PVC contents up to the administrator and play these recordings using the existing `asciinema play` application.
2. Add an interface to the ArgoCD control panel to view the list of recordings and embed a web player for these sessions: https://github.com/asciinema/asciinema-player.

**Comment by crenshaw-dev** [comment]
Prior art from the k8s ecosystem, doing basically the keystroke cache described above: https://github.com/Adyen/kubectl-rexec/blob/9accb3e53315128e40f4de362a083de1f18fcfc6/rexec/server/async.go#L16-L17

**Comment by loksonarius** [comment]
Hey all 👋

Came across this as a coworker's currently working on figuring out a way to audit exec sessions, and they actually proposed something similar to

> There is a project [asciinema](https://github.com/asciinema/asciinema) that tries to solve this very problem: record the entire terminal session, including all input, output, and timestamps of specific events, in a compact text file records.

They're looking at doing something like setting the shell to `asciinema rec` in the configuration of allowed shells ([relevant Argo docs](https://argo-cd.readthedocs.io/en/latest/operator-manual/web_based_terminal/#changing-allowed-shells)). We'll test this sometime soon and report on it if it's successful, but I wanted to leave a reply preemptively in case anyone else has tried similar, or if implementing asciinema on the Web terminal instead of depending on it being in the pod is worth considering for the project.

**Comment by cjreyn** [comment]
+1 for this feature

**Comment by n888** [comment]
Hi All,

EDIT: Proposal V2

I appreciate everyone's input and based off of the asciinema suggestions, please check out Proposal V2 for details and to test Terminal Session Recording with POC instructions:
- https://github.com/argoproj/argo-cd/issues/29251
